### PR TITLE
Align

### PR DESCRIPTION
**Pull request by MrAlpaca**

### CODE REVIEW

**Comment by MrAlpaca** [comment]
It looks like it should work. Can we run a test on this?

File: Robot Code/src/org/usfirst/frc/team3316/robot/vision/AlignShooter.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+		double turretAngle = Robot.turret.getAngle();
+		Robot.config.add("turret_Angle_SetPoint", turretAngle - towerAngle);
+	}
+	
+	private static void alignHood()
+	{
+		double distance = VisionServer.Data.get("DFC"); // DFC = Distance From Camera
+		// double hoodAngle = Robot.turret.getAngle();
+		double table [][] = null;
+		Robot.config.add("turret_Angle_SetPoint", Utils.valueInterpolation(distance, table));
```
Comment (karkason) [inline Robot Code/src/org/usfirst/frc/team3316/robot/vision/AlignShooter.java]: valueInterpolation throws an OutOfBoundsExecption with a null table. 
Provide a basic table.

**Comment by karkason** [comment]
How would it work if no target is seen? Does the python-side send that info? 
Do we need the consumers (HoodPID && TurretPID) to check for it? Is there a better way to write that up?

File: Robot Code/src/org/usfirst/frc/team3316/robot/vision/AlignShooter.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+		Robot.config.add("turret_Angle_SetPoint", turretAngle - towerAngle);
+	}
+
+	private static void alignHood()
+	{
+		double distance = VisionServer.Data.get("DFC"); // DFC = Distance From
+														// Camera
+		double table[][] = new double[][] { { 0 }, 
+											{ 0 } };
+		Robot.config.add("turret_Angle_SetPoint",
```
Comment (barakshelef) [inline Robot Code/src/org/usfirst/frc/team3316/robot/vision/AlignShooter.java]: This should be hood_Angle_SetPoint? or something like that...

File: Robot Code/src/org/usfirst/frc/team3316/robot/commands/hood/HoodBangbang.java [hunk 3]
```diff
@@ -1,40 +1,51 @@
...
 
-	protected void execute() {
+	protected void execute()
+	{
 		onVoltage = (double) config.get("hood_Bangbang_OnVoltage");
 		offVoltage = (double) config.get("hood_Bangbang_OffVoltage");
-		setPoint = (double) config.get("hood_Angle_SetPoint");
-		
-		if (Robot.hood.getAngle() <= setPoint) {
+		setPoint = (double) AlignShooter.alignHood();
```
Comment (karkason) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/hood/HoodBangbang.java]: Why is setPoint a global variable?

File: Robot Code/src/org/usfirst/frc/team3316/robot/commands/hood/HoodPID.java [hunk 4]
```diff
@@ -50,11 +53,11 @@ protected void init()
...
 		pid.setPID((double) config.get("hood_PID_KP"),
-					(double) config.get("hood_PID_KI"),
-					(double) config.get("hood_PID_KD"));
-		
-		pid.setSetpoint((double) config.get("hood_Angle_SetPoint"));
-		
+				(double) config.get("hood_PID_KI"),
+				(double) config.get("hood_PID_KD"));
+
+		pid.setSetpoint((double) AlignShooter.alignHood());
```
Comment (karkason) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/hood/HoodPID.java]: Move the alignHood call to a variable to mirror the way HoodBangbang does it.

File: Robot Code/src/org/usfirst/frc/team3316/robot/commands/turret/TurretBangbang.java [hunk 5]
```diff
@@ -1,40 +1,51 @@
...
 
-	protected void execute() {
+	protected void execute()
+	{
 		onVoltage = (double) config.get("turret_Bangbang_OnVoltage");
 		offVoltage = (double) config.get("turret_Bangbang_OffVoltage");
-		setPoint = (double) config.get("turret_Angle_SetPoint");
-		
-		if (Robot.turret.getAngle() <= setPoint) {
+		setPoint = (double) AlignShooter.alignTurret();
```
Comment (karkason) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/turret/TurretBangbang.java]: Same as the comment above about the global variable.

File: Robot Code/src/org/usfirst/frc/team3316/robot/commands/turret/TurretPID.java [hunk 6]
```diff
@@ -50,11 +53,11 @@ protected void init()
...
 		pid.setPID((double) config.get("turret_PID_KP"),
-					(double) config.get("turret_PID_KI"),
-					(double) config.get("turret_PID_KD"));
-		
-		pid.setSetpoint((double) config.get("turret_Angle_SetPoint"));
-		
+				(double) config.get("turret_PID_KI"),
+				(double) config.get("turret_PID_KD"));
+
+		pid.setSetpoint((double) AlignShooter.alignTurret());
```
Comment (karkason) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/turret/TurretPID.java]: Same as previous comment

File: Robot Code/src/org/usfirst/frc/team3316/robot/vision/VisionServer.java [hunk 7]
```diff
@@ -31,29 +35,42 @@
...
 			}
-			
-			String sentence = new String( receivePacket.getData() );
+
+			String sentence = new String(receivePacket.getData());
 			VisionServer.Data = parseLine(sentence);
+			if (VisionServer.Data.get("IDO") == 1.0)
+			{
+				AlignShooter.alignTurret();
+				AlignShooter.alignHood();
```
Comment (karkason) [inline Robot Code/src/org/usfirst/frc/team3316/robot/vision/VisionServer.java]: Why are these methods called both here and on PID/Bangbang?!

Comment (frc3316) [inline Robot Code/src/org/usfirst/frc/team3316/robot/vision/VisionServer.java]: \(*o*)/

File: Robot Code/src/org/usfirst/frc/team3316/robot/commands/hood/HoodPID.java [hunk 8]
```diff
@@ -38,7 +42,7 @@ public void pidWrite(double output)
 				pidOutput = output;
```
Comment (karkason) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/hood/HoodPID.java:42]: Consider moving the setMotors here, so we won't miss data.

**Comment by karkason** [comment]
Looking good!

**Pull request merged**